sw/device/tests/plic_sw_irq_test.c

```diff
@@ -0,0 +1,128 @@
+#include "rv_plic_regs.h"  // Generated.
+
+static const dif_rv_plic_target_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+
+static dif_rv_plic_t plic0;
+
+static bool external_intr_triggered = false;
+static bool software_intr_triggered = false;
+static bool software_intr_triggered_multiple = false;
+
+/**
+ * Configures all interrupts in PLIC.
+ */
+static void plic_configure_irqs(dif_rv_plic_t *plic) {
+  // Set IRQ priorities to MAX and Enable the IRQ
+  for (int i = 0; i < RV_PLIC_PARAM_NUM_SRC; ++i) {
+    CHECK_DIF_OK(dif_rv_plic_irq_set_priority(plic, i, kDifRvPlicMaxPriority));
+    CHECK_DIF_OK(
+        dif_rv_plic_irq_set_enabled(plic, i, kPlicTarget, kDifToggleEnabled));
+  }
+
+  // Set Ibex IRQ priority threshold level
+  CHECK_DIF_OK(dif_rv_plic_target_set_threshold(plic, kPlicTarget,
+                                                kDifRvPlicMinPriority));
+}
+
+/**
+ * External ISR.
+ *
+ * PLIC interrupt triggers ottf_external_isr() function. The expectation is
+ * MSIP triggers this. Inside the routine, it reads all Interrupt Pending bits
+ * and confirms only MSIP occurs.
+ */
+void ottf_external_isr(void) {
+  dif_rv_plic_irq_id_t interrupt_id;
+
+  external_intr_triggered = true;
+  LOG_FATAL("External Interrupt should not occur");
+  test_status_set(kTestStatusFailed);
+
+  // Clean up external ISRs (claim & complete)
+  CHECK_DIF_OK(dif_rv_plic_irq_claim(&plic0, kPlicTarget, &interrupt_id));
+  CHECK_DIF_OK(dif_rv_plic_irq_complete(&plic0, kPlicTarget, interrupt_id));
+}
+
+/**
+ * Software ISR
+ *
+ * Only SW ISR should occur in this test.
+ */
+void ottf_software_isr(void) {
+  ptrdiff_t msip_offset =
+      RV_PLIC_MSIP0_REG_OFFSET + (kPlicTarget * sizeof(uint32_t));
+  LOG_INFO("SOFTWARE ISR entered");
+  if (software_intr_triggered) {
+    // Should not occur multiple times
+    // Assume this is error
+    software_intr_triggered_multiple = true;
+  }
+  software_intr_triggered = true;
+
+  // Clean up software ISRs (claim? clear?)
+  // CHECK_DIF_OK(dif_rv_plic_software_irq_acknowledge(&plic0, kPlicTarget));
+  // Causing problems by calling functions above as similar as
+  // dif_rv_plic_software_irq_force(). Directly update MSIP
+  mmio_region_write32(plic0.base_addr, msip_offset, 0);
+
+  LOG_INFO("Software ISR event is cleared");
+}
+
+static void execute_test(dif_rv_plic_t *plic) {
+  // Force SW IRQ
+  // CHECK_DIF_OK(dif_rv_plic_software_irq_force(plic, kPlicTarget));
+  // Above returns kDifBadArg without reason. Run the function directly
+  ptrdiff_t msip_offset =
+      RV_PLIC_MSIP0_REG_OFFSET + (kPlicTarget * sizeof(uint32_t));
+  mmio_region_write32(plic->base_addr, msip_offset, 1);
+
+  // Need gap to correctly return. Refer lowRISC/opentitan#12785
+  __asm__("nop\n\tnop\n\t");
+
+  // Wait SW IRQ being processed
+  // LOG_INFO("Entering the while (software_intr_triggered == false)");
+  while (abs_mmio_read32((uint32_t)&software_intr_triggered) == false) {
```

Comment:
The size of a `bool` is 1 byte in our implementation (Side note: the size of a `bool` is implementation-defined and not given as a particular value by the C standard.). Shall we just make the globals be `uint32_t` and use 0 / nonzero for false / true?

Reply:
Revised